pkg/envoy/xdsutil.go

```diff
@@ -260,18 +243,50 @@ func MessageToAny(pb proto.Message) (*any.Any, error) {
 }
 
 // GetDownstreamTLSContext creates a downstream Envoy TLS Context
-func GetDownstreamTLSContext(serviceName service.MeshService, mTLS bool) *xds_auth.DownstreamTlsContext {
+func GetDownstreamTLSContext(upstreamSvc service.MeshService, mTLS bool) *xds_auth.DownstreamTlsContext {
+	upstreamSDSCert := SDSCert{
+		MeshService: upstreamSvc,
+		CertType:    ServiceCertType,
+	}
+
+	var downstreamPeerValidationCertType SDSCertType
+	if mTLS {
+		// Perform SAN validation for downstream client certificates
+		downstreamPeerValidationCertType = RootCertTypeForMTLSInbound
+	} else {
+		// TLS based cert validation (used for ingress)
+		downstreamPeerValidationCertType = RootCertTypeForHTTPS
+	}
+	// The downstream peer validation SDS cert points to a cert with the name 'upstreamSvc' only
+	// because we use a single DownstreamTlsContext for all inbound traffic to the given 'upstreamSvc'.
+	// This single DownstreamTlsContext is used to validate all allowed inbound SANs with the
+	// 'RootCertTypeForMTLSInbound' cert type used for in-mesh downstreams, while 'RootCertTypeForHTTPS'
+	// cert type is used for non-mesh downstreams such as ingress.
+	downstreamPeerValidationSDSCert := SDSCert{
+		MeshService: upstreamSvc,
+		CertType:    downstreamPeerValidationCertType,
+	}
+
 	tlsConfig := &xds_auth.DownstreamTlsContext{
-		CommonTlsContext: getCommonTLSContext(serviceName, mTLS, Inbound),
+		CommonTlsContext: getCommonTLSContext(upstreamSDSCert, downstreamPeerValidationSDSCert),
 		// When RequireClientCertificate is enabled trusted CA certs must be provided via ValidationContextType
 		RequireClientCertificate: &wrappers.BoolValue{Value: mTLS},
 	}
 	return tlsConfig
 }
 
-// GetUpstreamTLSContext creates an upstream Envoy TLS Context
-func GetUpstreamTLSContext(serviceName service.MeshService, sni string) *xds_auth.UpstreamTlsContext {
-	commonTLSContext := getCommonTLSContext(serviceName, true /* mTLS */, Outbound)
+// GetUpstreamTLSContext creates an upstream Envoy TLS Context for the given downstream and upstream service pair
+func GetUpstreamTLSContext(downstreamSvc, upstreamSvc service.MeshService) *xds_auth.UpstreamTlsContext {
```

Comment:
```suggestion
func GetUpstreamTLSContext(downstreamSvc service.MeshService, upstreamSvc service.MeshService) *xds_auth.UpstreamTlsContext {
```

Reply:
I'll defer addressing this since this is just Go syntax for shorthand function signature.